### ISSUE TITLE
[k8s] Support EC2 GPU instance types

### ISSUE THREAD

**Issue by jhaynes** [bottlerocket-os/bottlerocket#769]
**What I'd like:**

Allow customers to enable GPU drivers at runtime somehow...

**Any alternatives you've considered:**

The alternative of building a separate AMI is an option, but not nearly as desirable as a single AMI with a runtime toggle to enable the required drivers based on instance type.

**Comment by kniec** [comment]
Do you foresee this to make a clear cut between what the host system needs to run (kernel drivers) and what the container should carry (user land driver)? As of today a lot of times the assumption is that the container will just bind-mount the user-land driver from the host, hope for the best (which is usually an ok assumption) and replace whatever cuda was insight the container.
Since CUDA is rather backwards compatible (what I've heard) it might even be possible to only install the latest and greatest kernel driver on the host and force the container to carry the user-land driver and stop being lazy by putting the burden on the host installation...

**Comment by jhaynes** [comment]
We haven't decided on the best path forward for GPUs in Bottlerocket but are looking at a few options. The one we're most excited about is modeling our support similarly to the [NVIDIA Operator](https://github.com/NVIDIA/gpu-operator). That way, customers only have to worry about one AMI or disk image and install the appropriate drivers at runtime via containers.

As far as providing the userland drivers vs just kernel, that is another decision point that we haven't made yet. However, I don't expect us to ask customers to dramatically re-think how they're using GPUs just to be able to use Bottlerocket.

**Comment by mozinrat** [comment]
Just wanted to understand is there any formal approach decided for this? Its been quiet some time since any updates.

**Comment by samuelkarp** [comment]
Hi @mozinrat! The last update provided by @jhaynes is still accurate; we're still planning to look at this but we haven't yet determined a particular approach.

**Comment by rverma-dev** [comment]
Is it possible to address https://github.com/bottlerocket-os/bottlerocket/issues/1031 too along with this. With this we can focus better for ML workloads on bottlerocket.

**Comment by awoimbee** [comment]
Is this issue solved ? 
https://github.com/weaveworks/eksctl/issues/4760
> Bottlerocket is releasing an AMI specifically supporting NVIDIA based GPU EC2 instance types

**Comment by sodabrew** [comment]
Is this the blog post?
https://aws.amazon.com/about-aws/whats-new/2022/03/bottlerocket-support-gpu-ec2-instance-types-powered-by-nvidia/

https://aws.amazon.com/blogs/containers/bottlerocket-support-for-nvidia-gpus/

Does this support divvying out GPUs to containers, for example a g4dn.12xlarge has 4x NVIDIA T4 GPUs. Can I bind one to each of four containers?

**Comment by arnaldo2792** [comment]
Hello folks! Sorry for the late response:

@awoimbee, yes that issue is resolved and you can use `eksctl` with self-managed nodegroups, with x86_64 hosts that include NVIDIA GPUs. There is still some work going on to support managed nodegroups and Graviton instances.

@sodabrew Correct, those are the blogposts. And yes, you can use a different GPU per container, in instances that have multiple GPUs. As an example, I deployed one pod with 2 containers in it, each with a different GPU attached to them:

```bash
k8s ❯ kubectl exec gpu-tests-8jx8m -c gpu-tests -- bash -c 'echo $NVIDIA_VISIBLE_DEVICES'
GPU-4f7b097f-33f2-335e-0d5c-7f4c9ffed49b

k8s ❯ kubectl exec gpu-tests-8jx8m -c gpu-tests-2 -- bash -c 'echo $NVIDIA_VISIBLE_DEVICES'
GPU-e4eec57c-ce0a-3db9-c557-0b886404e686
```

This is the configuration file that I used:

```yaml
apiVersion: apps/v1
kind: DaemonSet
metadata:
  name: gpu-tests
spec:
  selector:
    matchLabels:
      name: gpu-tests
  template:
    metadata:
      labels:
        name: gpu-tests
    spec:
      containers:
        - name: gpu-tests
          image: amazonlinux:2
          command: ['sh', '-c', 'sleep infinity']
          resources:
            limits:
               nvidia.com/gpu: 1
          env:
            - name: NVIDIA_DRIVER_CAPABILITIES
              value: all
        - name: gpu-tests-2
          image: amazonlinux:2
          command: ['sh', '-c', 'sleep infinity']
          resources:
            limits:
               nvidia.com/gpu: 1
          env:
            - name: NVIDIA_DRIVER_CAPABILITIES
              value: all

```

**Comment by sodabrew** [comment]
Thank you! Oh, also is GPU partitioning supported at this time?

**Comment by arnaldo2792** [comment]
Hey @sodabrew, for the moment we don't support [MIG](https://docs.nvidia.com/datacenter/cloud-native/kubernetes/mig-k8s.html), but it is part of our roadmap.

**Comment by arnaldo2792** [comment]
We released this on 1.6.0. Refer to [the main documentation](https://github.com/bottlerocket-os/bottlerocket#nvidia-gpus-support) to read about how to use NVIDIA GPUs with Bottlerocket
